src/Control/Distributed/Spark/Iterator.hs

```diff
@@ -0,0 +1,102 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Control.Distributed.Spark.Iterator
+  ( Iterator(..)
+  , JIterator
+  , emptyIterator
+  , foldIterator
+  , singletonIterator
+  ) where
+
+import Control.Distributed.Closure.TH
+import Data.Int
+import Data.IORef
+import Foreign.JNI
+import Foreign.Ptr
+import Foreign.StablePtr
+import Language.Java
+import System.IO (fixIO)
+
+
+-- | This is a haskell representation for iterators.
+newtype Iterator a = Iterator (IO (Maybe (a, Iterator a)))
```

Comment:
Why are we defining our own iterator type (which incidentally can't participate in fusion, and also is not usable without hoisting any functions over them into IO yet `RDD.map` only supports pure functions) instead of reusing an existing (and de facto standard) type like `Conduit`, or perhaps `Pipe`?